design/cors-configuration.md

```diff
@@ -0,0 +1,188 @@
+}
+
+// CorsPolicy allows for setting de CORS policy
+type CorsPolicy struct {
+    // Specifies whether the resource allows credentials.
+    AllowCredentials bool `json:"allowCredentials"`
+    // AllowOrigin specifies the origins that will be allowed to do CORS requests.
+    AllowOrigin []string `json:"allowOrigin"`
+    // AllowMethods specifies the content for the *access-control-allow-methods* header.
+    AllowMethods []string `json:"allowMethods"`
+    // AllowHeaders specifies the content for the *access-control-allow-headers* header.
+    AllowHeaders []string `json:"allowHeaders"`
+    // ExposeHeaders Specifies the content for the *access-control-expose-headers* header.
+    ExposeHeaders []string `json:"exposeHeaders"`
+    // MaxAge specifies the content for the *access-control-max-age* header.
+    MaxAge string `json:"maxAge"`
+}
+```
+
+The common DAG struct will be updated accordingly:
+
+```go
+// contour/internal/dag/dag.go
+type Route struct {
+   Prefix       string
+    object       interface{} // one of Ingress or IngressRoute
+   httpServices map[servicemeta]*HTTPService
+  
+   [... other members ...]
+    // Indicates the CORS policy to apply
+    CorsPolicy *ingressroutev1.CorsPolicy
+}
+```
+
+### Communicating from DAG to Envoy
+
+Once the DAG has been built, we need to send the CORS related configuration to Envoy. To do so, we will update the `RouteRoute` function of `contour/internal/envoy/route.go` and map the values from DAG's route to [protobuf](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html?highlight=shadow#route-corspolicy).
+
+One important thing to take into account is that in Envoy 1.10 the [corsPolicy.enabled has been deprecated](https://www.envoyproxy.io/docs/envoy/v1.10.0/api-v2/api/v2/route/route.proto#envoy-api-field-route-corspolicy-enabled). As we are still using Envoy 1.9, we'll start using the deprecated field to enable CORS and change it once we upgrade to 1.10.
+
+### Enabling the CORS filter in Envoy
+
+The final step will be to enable the Envoy's CORS filter. This we'll be done in the `HTTPConnectionManager` function of `contour/internal/listener.go`.
+
+## Alternatives Considered
+
+### Delegate all the CORS related logic to the applications
+All the CORS related tasks could be handled in each application using an application level middleware or similar. However, it isn't something people usually want to do at this level for the following reasons: 
+
+- This kind of configuration doesn't usually rely on developers because they are mainly focused on developing the services (the what), and not on which domains those services are going to be deployed on (the where), nor which security rules should be applied. This is usually a task for the Ops/DevOps/DevSecOps teams and that's the reason why all the reverse proxies out there provide this functionality alongside SSL termination, gzip compression and so on.
```

Comment:
Perhaps this argues for specifying the CORS policy somewhere that is not a `Route`? The inclusion model for `HTTPProxy` is more or less that cluster or infrastructure operators set up the `HTTPProxy` and some sort of inclusion chain, then application developers configure the `Route` objects.

It seems like here you are arguing that CORS policy should be centralized rather than distributed.

Maybe the trade-off here is to centralize the policy, but allow it to be disabled on specific routes?

Reply:
> Maybe the trade-off here is to centralize the policy, but allow it to be disabled on specific routes?

Sounds good to me.

Reply:
We've done similar things before (a centralized policy that can be disabled), I also like this pattern.

Reply:
We'll start setting the policy at `HTTPProxy` level and see how things go. We could add a property at route level to disable the policy in following PRs.